File: communication/mosquitto/simple-demo/python-server/mosquitto_python_server/consumer/listener.py
import base64
import json
import logging

from mosquitto_python_server.common.types import PayloadMessage
from mosquitto_python_server.common.values import TOPIC_PREFIX, QOS
from mosquitto_python_server.config.mqtt import MQTT

from paho.mqtt.client import MQTTMessage, Client


class Listener:
    def __init__(self, mqtt: MQTT):
        self.mqtt = mqtt

    def start_consuming(self):
        def on_message(_: Client, userdata, msg: MQTTMessage):
            payload = json.loads(msg.payload.decode())
            message = PayloadMessage(**payload)
            logging.info(f"Received Message: {message}")
            logging.info(f"Received Message Data: {base64.b64decode(message.data).decode('utf-8')}")
            logging.info(f"Received Topic: {msg.topic}")

        self.mqtt.client.on_message = on_message
        self.mqtt.client.subscribe(TOPIC_PREFIX + "#", QOS)

File: communication/mosquitto/simple-demo/go-server/consumer/listener.go
package consumer

import (
	"encoding/json"
	"log"

	mqtt "github.com/eclipse/paho.mqtt.golang"

	"github.com/seungyeop-lee/poc/communication/mosquitto/simple-demo/go-server/common"
	"github.com/seungyeop-lee/poc/communication/mosquitto/simple-demo/go-server/config"
)

type Listener struct {
	mqtt *config.MQTT
}

func NewListener(mqtt *config.MQTT) *Listener {
	return &Listener{mqtt: mqtt}
}

func (l *Listener) StartConsuming() error {
	topic := common.TopicPrefix + "#"
	token := l.mqtt.Client.Subscribe(topic, common.QoS, func(client mqtt.Client, msg mqtt.Message) {
		var payload common.PayloadMessage
		_ = json.Unmarshal(msg.Payload(), &payload)
		log.Printf("Received Message: %+v\n", payload)
		log.Printf("Received Message Data: %s\n", string(payload.Data))
		log.Printf("Received Topic: %s\n", msg.Topic())
	})

	if token.Wait() && token.Error() != nil {
		return token.Error()
	}

	return nil
}
